### ISSUE TITLE
no_std with customized panic handler

### ISSUE THREAD

**Issue by dingelish** [CraneStation/cranelift#697]
Hi there,

cranelift is really amazing and thanks for your hard working!

I'm trying to use cranelift-codegen in a no_std environment with a customized panic handler. The codes are like this:

```
#![no_std]
#![feature(lang_items)]
use core::panic::PanicInfo;

// Not work if uncommented
//use cranelift_codegen;

#[panic_handler]
fn panic(_info: &PanicInfo) -> ! {
    unimplemented!()
}

#[lang = "eh_personality"]
extern fn eh_personality() {}
```

In Cargo.toml:
```
cranelift-codegen = { version = "0.29", default-features = false, features = ["core"] }
```

Errors are like

```
error[E0152]: duplicate lang item found: `panic_impl`.
  --> src/lib.rs:8:1
   |
8  | / fn panic(_info: &PanicInfo) -> ! {
9  | |     unimplemented!()
10 | | }
   | |_^
   |
   = note: first defined in crate `std`.

error[E0152]: duplicate lang item found: `eh_personality`.
  --> src/lib.rs:13:1
   |
13 | extern fn eh_personality() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: first defined in crate `panic_unwind`.

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0152`.
```

These errors make me believe that something brings `std` into this crate somewhere. But I can't find  it (not familiar with cranelift). Could you please help me with it? Thanks!

**Comment by sunfishcode** [comment]
I just created an empty project and copied the code here into it, and wasn't able to reproduce the error. Is it possible that you have some other dependency which is pulling in `std`?

[cargo nono](https://github.com/hobofan/cargo-nono) is a tool for testing `no_std` support which can sometimes help debug these kinds of issues.

**Comment by dingelish** [comment]
Hi @sunfishcode , I just uploaded the codes to this repo
https://github.com/dingelish/cranelift-codegen-nostd-test

Could you please try it? Thanks! cargo-nono is panicking on this case after outputs `cg: SUCCESS`.

**Comment by jgouly** [comment]
I cloned the repo, and can reproduce.

**Comment by sunfishcode** [comment]
I can reproduce it too. I don't yet know what's going on, but some possibilities are:
 - Cranelift is still using `hashmap_core` in no_std builds, which is no longer maintained; we should migrate to `hashbrown`.
 - I just published a target-lexicon to 0.3.0 (and faerie 0.9.0 to accompany it), which upgrade target-lexicon to Rust 2018 Edition, which involves changes the way no_std packages work. We should upgrade cranelift to these versions and see if that helps.

**Comment by jgouly** [comment]
I'm pretty sure it's this bug: https://github.com/rust-lang/cargo/issues/4866

This fixes it:
    diff --git a/cranelift-codegen/meta/Cargo.toml b/cranelift-codegen/meta/Cargo.toml
    index 1253c89e..31aebd25 100644
    --- a/cranelift-codegen/meta/Cargo.toml
    +++ b/cranelift-codegen/meta/Cargo.toml
    @@ -9,7 +9,7 @@ readme = "README.md"
     edition = "2018"
    
     [dependencies]
    -cranelift-entity = { path = "../../cranelift-entity", version = "0.29.0" }
    +cranelift-entity = { path = "../../cranelift-entity", version = "0.29.0", default-features = false }
    
     [badges]
     maintenance = { status = "experimental" }

**Comment by sunfishcode** [comment]
Thanks @jgouly for figuring that out! I think I have a workaround for it here: https://github.com/CraneStation/cranelift/pull/699/

**Comment by dingelish** [comment]
Thanks @jgouly !

**Comment by sunfishcode** [comment]
This is now fixed in #699.
